public/app/core/history/richHistoryLocalStorage.ts

```diff
@@ -0,0 +1,172 @@
+import RichHistoryStorage, {
+  RichHistoryServiceError,
+  RichHistoryStorageWarning,
+  RichHistoryStorageWarningDetails,
+} from './richHistoryStorage';
+import { RichHistoryQuery } from '../../types';
+import store from '../store';
+import { DataQuery } from '@grafana/data';
+import { isEqual, omit } from 'lodash';
+import { createRetentionPeriodBoundary, RICH_HISTORY_SETTING_KEYS } from './richHistoryLocalStorageUtils';
+
+export const RICH_HISTORY_KEY = 'grafana.explore.richHistory';
+export const MAX_HISTORY_ITEMS = 10000;
+
+/**
+ * Local storage implementation for Rich History. It keeps all entries in browser's local storage.
+ */
+export default class RichHistoryLocalStorage implements RichHistoryStorage {
+  /**
+   * Return all history entries, perform migration and clean up entries not matching retention policy.
+   */
+  async getRichHistory(): Promise<RichHistoryQuery[]> {
+    const richHistory: RichHistoryQuery[] = store.getObject(RICH_HISTORY_KEY, []);
+    const transformedRichHistory = migrateRichHistory(richHistory);
+    return transformedRichHistory;
+  }
+
+  async addToRichHistory(richHistoryQuery: RichHistoryQuery): Promise<RichHistoryStorageWarningDetails | undefined> {
+    let richHistory = cleanUp(await this.getRichHistory());
+
+    /* Compare queries of a new query and last saved queries. If they are the same, (except selected properties,
+     * which can be different) don't save it in rich history.
+     */
+    const newQueriesToCompare = richHistoryQuery.queries.map((q) => omit(q, ['key', 'refId']));
+    const lastQueriesToCompare =
+      richHistory.length > 0 &&
+      richHistory[0].queries.map((q) => {
+        return omit(q, ['key', 'refId']);
+      });
+
+    if (isEqual(newQueriesToCompare, lastQueriesToCompare)) {
+      const error = new Error('Entry already exists');
+      error.name = RichHistoryServiceError.DuplicatedEntry;
+      throw error;
+    }
+
+    const { queriesToKeep, limitExceeded } = checkLimits(richHistory);
+
+    let updatedHistory: RichHistoryQuery[] = [richHistoryQuery, ...queriesToKeep];
+
+    try {
+      store.setObject(RICH_HISTORY_KEY, updatedHistory);
+    } catch (error) {
+      if (error.name === 'QuotaExceededError') {
+        throwError(RichHistoryServiceError.StorageFull, `Saving rich history failed: ${error.message}`);
+      } else {
+        throw error;
+      }
+    }
+
+    if (limitExceeded) {
+      return {
+        type: RichHistoryStorageWarning.LimitExceeded,
+        message: `Query history reached the limit of ${MAX_HISTORY_ITEMS}. Old, not-starred items have been removed.`,
+      };
+    } else {
+      return undefined;
+    }
```

Comment:
nit: it may read better without the else statement
```suggestion
```

Reply:
Updated in 84e714bbd151e7416752e1acd7c7c42867a3d82c